li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaFederatedProducerImpl.java

```diff
@@ -208,7 +230,85 @@ private LiKafkaFederatedProducerImpl(LiKafkaProducerConfig configs,
+      Integer partition = producerRecord.partition();
+      Future<RecordMetadata> future = null;
+      UUID messageId = _uuidFactory.getUuid(producerRecord);
+      if (LOG.isTraceEnabled()) {
+        LOG.trace("Sending event: [{}, {}] with key {} to kafka topic {} to cluster {}",
+            messageId.toString().replaceAll("-", ""),
+            value.toString(),
+            (key != null) ? key.toString() : "[none]",
+            topic,
+            context.cluster().name());
+      }
+      byte[] serializedValue;
+      byte[] serializedKey;
+      try {
+        serializedValue = _valueSerializer.serialize(topic, value);
+        serializedKey = _keySerializer.serialize(topic, key);
+      } catch (Throwable t) {
+        // Audit the attempt and the failure.
+        auditor.record(auditToken, topic, timestamp, 1L, 0L, AuditType.ATTEMPT);
+        throw t;
+      }
+      int sizeInBytes = (serializedKey == null ? 0 : serializedKey.length)
+          + (serializedValue == null ? 0 : serializedValue.length);
+      // Audit the attempt.
+      auditor.record(auditToken, topic, timestamp, 1L, (long) sizeInBytes, AuditType.ATTEMPT);
+      // We wrap the user callback for error logging and auditing purpose.
+      Callback errorLoggingCallback =
+          new ErrorLoggingCallback<>(messageId, auditToken, topic, timestamp, sizeInBytes, auditor, callback);
+      if (_largeMessageEnabled && serializedValue != null && serializedValue.length > _maxMessageSegmentSize) {
+        List<ProducerRecord<byte[], byte[]>> segmentRecords =
+            _messageSplitter.split(topic, partition, timestamp, messageId, serializedKey, serializedValue);
+        Callback largeMessageCallback = new LargeMessageCallback(segmentRecords.size(), errorLoggingCallback);
+        for (ProducerRecord<byte[], byte[]> segmentRecord : segmentRecords) {
+          future = producer.send(segmentRecord, largeMessageCallback);
+        }
+      } else {
+        // In order to make sure consumer can consume both large message segment and the ordinary message,
+        // we wrap the normal message as a single segment large message. When consumer sees it, it will
+        // be returned by message assembler immediately. We set a pretty large maxSegmentSize to make sure
+        // the message will end up in one segment.
+        List<ProducerRecord<byte[], byte[]>> wrappedRecord =
+            _messageSplitter.split(topic, partition, timestamp, messageId, serializedKey, serializedValue, Integer.MAX_VALUE / 2);
+        assert (wrappedRecord.size() == 1);
+        future = producer.send(wrappedRecord.get(0), errorLoggingCallback);
+      }
+      failed = false;
+      return future;
+    } catch (SkippableException e) {
+      LOG.warn("Exception thrown when producing message to partition {}-{}", producerRecord.topic(), producerRecord.partition());
+      return null;
+    } finally {
+      if (failed) {
```

Comment:
the auditing bit should come last - after all the counter updates have happened (in case the user-provided auditor crashes)

Reply:
done